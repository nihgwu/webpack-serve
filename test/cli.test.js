const { join, resolve } = require('path');

const execa = require('execa');
const request = require('supertest');

const cliPath = resolve(__dirname, '../lib/cli.js');

// eslint-disable-next-line no-unused-vars
const pipe = (proc) => {
  proc.stderr.pipe(process.stdout);
  proc.stdout.pipe(process.stdout);
};

const run = (flags, execaOptions = {}) => {
  const args = [cliPath, flags, execaOptions];
  const reReady = /(Compiled successfully)|(Compiled with warnings)/i;
  const proc = execa(...args);

  // pipe(proc);

  proc.ready = new Promise((resolv) => {
    proc.stdout.on('data', (data) => {
      if (reReady.test(data.toString())) {
        resolv();
      }
    });
  });

  return proc;
};

jest.setTimeout(10000);

describe('cli', () => {
  test('--help', () => {
    const proc = run(['--help']);

    return proc.then(({ stdout }) => {
      expect(stdout).toMatchSnapshot();
    });
  });

  test('--version', () => {
    const proc = run(['--version']);

    return proc.then(({ stdout }) => {
      expect(stdout).toMatch(/\d+\.\d+\.\d+/);
    });
  });

  test('--config', () => {
    const config = './test/fixtures/basic/webpack.config.js';
    const proc = run(['--config', config, '--port', '8888']);

    return proc.ready.then(() =>
      request('http://localhost:8888')
        .get('/output.js')
        .expect(200)
        .then(() => proc.kill('SIGINT'))
    );
  });

  test('bad config', () => {
    const config = './test/fixtures/invalid.config.js';
    const proc = run(['--config', config]);

    return proc.catch((e) => {
      const message = e.message
        .split(/\n\s+at/)[0]
        .replace(
          /\(node:\d+\) ExperimentalWarning: The fs.promises API is experimental\n/,
          ''
        );
      expect(message).toMatchSnapshot();
      expect(proc.exitCode).toBe(1);
    });
  });

  test('config throwing error', () => {
    const config = './test/fixtures/error.config.js';
    const proc = run(['--config', config]);

    return proc.catch((e) => {
      const message = e.message.split(/\n\s+at/);
      expect(message[0]).toMatchSnapshot();
      expect(proc.exitCode).toBe(1);
    });
  });

  test('serve.config.js', () => {
    const config = './basic/webpack.config.js';
    const cwd = join(__dirname, 'fixtures');
    const proc = run(['--config', config], { cwd });

    return proc.ready.then(() =>
      request('http://0.0.0.0:8080')
        .get('/output.js')
        .expect(200)
        .then(() => proc.kill('SIGINT'))
    );
  });

  test('[config]', () => {
    const serveConfig = './test/fixtures/basic/custom-serve.config.js';
    const config = './test/fixtures/basic/webpack.config.js';
    const proc = run([serveConfig, '--config', config]);

    return proc.ready.then(() =>
      request('http://0.0.0.0:8888')
        .get('/output.js')
        .expect(200)
        .then(() => proc.kill('SIGINT'))
    );
  });
});
