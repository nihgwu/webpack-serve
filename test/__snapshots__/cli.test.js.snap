// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli --help 1`] = `
"
  A lean, modern, and flexible webpack development server

  Usage
    $ webpack-serve <config> [...options]

  <config>
    Custom serve config file instead of default search path.

  Options
    --clipboard      Specify whether or not the server should copy the server URI to the clipboard (default: true)
    --config         The webpack config to serve.
    --content        The path from which static content will be served (default: process.cwd)
    --dev-ware       Set options for webpack-dev-middleware. e.g. --dev-ware.publicPath /
    --help           Show usage information and the options listed here.
    --host           The host the app should bind to
    --hot-client     Set options for webpack-hot-client. e.g. --hot-client.host localhost
                     Use --no-hot-client to disable webpack-hot-client
    --http2          Instruct the server to use HTTP2
    --https-cert     Specify a filesystem path of an SSL certificate. Must be paired with a key
    --https-key      Specify a filesystem path of an SSL key. Must be paired with a cert
    --https-pass     Specify a passphrase to enable https. Must be paired with a pfx file
    --https-pfx      Specify a filesystem path of an SSL pfx file. Must be paired with a passphrase
    --log-level      Limit all process console messages to a specific level and above
                     Levels: trace, debug, info, warn, error, silent
    --log-time       Instruct the logger for webpack-serve and dependencies to display a timestamp
    --hmr            Specify whether or not the client should apply Hot Module Replacement patches (default: true)
    --reload         Specify whether or not the middleware should reload the page for build errors (default: true)
    --open           Instruct the app to open in the default browser
    --open-app       The name of the app to open the app within, or an array
                     containing the app name and arguments for the app
    --open-path      The path with the app a browser should open to
    --port           The port the app should listen on. Default: 8080
    --require, -r    Preload one or more modules before loading the webpack configuration
    --version        Display the webpack-command version

  Note: Any boolean flag can be prefixed with 'no-' instead of specifying a value.
  e.g. --no-reload rather than --reload=false

  Examples
    $ webpack-serve ./webpack.config.js --no-reload
    $ webpack-serve --config ./webpack.config.js --port 1337
    $ webpack-serve # config can be omitted for webpack v4+ only
"
`;

exports[`cli bad config 1`] = `
"Command failed: <PROJECT_ROOT>/lib/cli.js --config ./test/fixtures/invalid.config.js

ValidationError: config-loader

  Options Validation Error

  options['foo']  is an invalid additional property"
`;

exports[`cli config throwing error 1`] = `
"Command failed: <PROJECT_ROOT>/lib/cli.js --config ./test/fixtures/error.config.js
<PROJECT_ROOT>/test/fixtures/error.config.js:3
  throw new Error('batman');
  ^

Error: batman"
`;
